### PR TITLE
refactor: Move from Gas to NearGas from near-gas crate

### PR DESCRIPTION
**Pull request by iho**

### CODE REVIEW

File: examples/factory-contract/low-level/src/lib.rs [hunk 1]
```diff
@@ -4,7 +4,7 @@ use near_sdk::serde_json;
 use near_sdk::{env, near_bindgen, AccountId, Gas, PromiseResult};
 
 // Prepaid gas for making a single simple call.
-const SINGLE_CALL_GAS: Gas = Gas(20_000_000_000_000);
+const SINGLE_CALL_GAS: Gas = Gas::from_ggas(20);
```
Comment (frol) [inline examples/factory-contract/low-level/src/lib.rs]:
```suggestion
const SINGLE_CALL_GAS: Gas = Gas::from_tgas(20);
```

File: examples/fungible-token/test-contract-defi/src/lib.rs [hunk 2]
```diff
@@ -10,7 +10,7 @@ use near_sdk::{
 
 const BASE_GAS: u64 = 5_000_000_000_000;
 const PROMISE_CALL: u64 = 5_000_000_000_000;
-const GAS_FOR_FT_ON_TRANSFER: Gas = Gas(BASE_GAS + PROMISE_CALL);
+const GAS_FOR_FT_ON_TRANSFER: Gas = Gas::from_gas(BASE_GAS + PROMISE_CALL);
```
Comment (frol) [inline examples/fungible-token/test-contract-defi/src/lib.rs:13]:
```rust
const BASE_GAS: Gas = Gas::from_tgas(5);
const PROMISE_CALL: Gas = Gas::from_tgas(5);
const GAS_FOR_FT_ON_TRANSFER: Gas = Gas::from_gas(BASE_GAS.as_gas() + PROMISE_CALL.as_gas());
```

File: examples/non-fungible-token/test-approval-receiver/src/lib.rs [hunk 3]
```diff
@@ -4,14 +4,11 @@ A stub contract that implements nft_on_approve for e2e testing nft_approve.
...
-use near_sdk::{
-    env, log, near_bindgen, require, AccountId, Gas, PanicOnDefault,
-    PromiseOrValue,
-};
+use near_sdk::{env, log, near_bindgen, require, AccountId, Gas, PanicOnDefault, PromiseOrValue};
 
 const BASE_GAS: u64 = 5_000_000_000_000;
 const PROMISE_CALL: u64 = 5_000_000_000_000;
-const GAS_FOR_NFT_ON_APPROVE: Gas = Gas(BASE_GAS + PROMISE_CALL);
+const GAS_FOR_NFT_ON_APPROVE: Gas = Gas::from_gas(BASE_GAS + PROMISE_CALL);
```
Comment (frol) [inline examples/non-fungible-token/test-approval-receiver/src/lib.rs:11]: Apply the same as above.

File: examples/non-fungible-token/test-token-receiver/src/lib.rs [hunk 4]
```diff
@@ -4,13 +4,11 @@ A stub contract that implements nft_on_transfer for simulation testing nft_trans
...
 use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
-use near_sdk::{
-    env, log, near_bindgen, require, AccountId, Gas, PanicOnDefault, PromiseOrValue,
-};
+use near_sdk::{env, log, near_bindgen, require, AccountId, Gas, PanicOnDefault, PromiseOrValue};
 
 const BASE_GAS: u64 = 5_000_000_000_000;
 const PROMISE_CALL: u64 = 5_000_000_000_000;
-const GAS_FOR_NFT_ON_TRANSFER: Gas = Gas(BASE_GAS + PROMISE_CALL);
+const GAS_FOR_NFT_ON_TRANSFER: Gas = Gas::from_gas(BASE_GAS + PROMISE_CALL);
```
Comment (frol) [inline examples/non-fungible-token/test-token-receiver/src/lib.rs:11]: ^

File: near-contract-standards/src/fungible_token/core_impl.rs [hunk 5]
```diff
@@ -10,8 +10,8 @@ use near_sdk::{
     PromiseResult, StorageUsage,
 };
 
-const GAS_FOR_RESOLVE_TRANSFER: Gas = Gas(5_000_000_000_000);
-const GAS_FOR_FT_TRANSFER_CALL: Gas = Gas(25_000_000_000_000 + GAS_FOR_RESOLVE_TRANSFER.0);
+const GAS_FOR_RESOLVE_TRANSFER: Gas = Gas::from_ggas(5);
+const GAS_FOR_FT_TRANSFER_CALL: Gas = Gas::from_ggas(30);
```
Comment (frol) [inline near-contract-standards/src/fungible_token/core_impl.rs]:
```suggestion
const GAS_FOR_RESOLVE_TRANSFER: Gas = Gas::from_tgas(5);
const GAS_FOR_FT_TRANSFER_CALL: Gas = Gas::from_tgas(30);
```

File: near-contract-standards/src/fungible_token/core_impl.rs [hunk 6]
```diff
@@ -133,17 +133,20 @@ impl FungibleTokenCore for FungibleToken {
         msg: String,
     ) -> PromiseOrValue<U128> {
         assert_one_yocto();
-        require!(env::prepaid_gas() > GAS_FOR_FT_TRANSFER_CALL, "More gas is required");
+        require!(
+            env::prepaid_gas().as_gas() > GAS_FOR_FT_TRANSFER_CALL.as_gas(),
```
Comment (frol) [inline near-contract-standards/src/fungible_token/core_impl.rs:137]: As you already implemented Ord and PartialOrd for NearGas in your PR, you can revert this change back to:

```rust
env::prepaid_gas() > GAS_FOR_FT_TRANSFER_CALL,
```

File: near-contract-standards/src/fungible_token/core_impl.rs [hunk 7]
```diff
@@ -133,17 +133,20 @@ impl FungibleTokenCore for FungibleToken {
...
+            "More gas is required"
+        );
         let sender_id = env::predecessor_account_id();
         let amount: Balance = amount.into();
         self.internal_transfer(&sender_id, &receiver_id, amount, memo);
         let receiver_gas = env::prepaid_gas()
-            .0
-            .checked_sub(GAS_FOR_FT_TRANSFER_CALL.0)
+            .as_gas()
+            .checked_sub(GAS_FOR_FT_TRANSFER_CALL.as_gas())
```
Comment (frol) [inline near-contract-standards/src/fungible_token/core_impl.rs:145]:
```suggestion
            .checked_sub(GAS_FOR_FT_TRANSFER_CALL)
```

File: near-contract-standards/src/fungible_token/core_impl.rs [hunk 8]
```diff
@@ -133,17 +133,20 @@ impl FungibleTokenCore for FungibleToken {
...
             .unwrap_or_else(|| env::panic_str("Prepaid gas overflow"));
         // Initiating receiver's call and the callback
         ext_ft_receiver::ext(receiver_id.clone())
-            .with_static_gas(receiver_gas.into())
+            .with_static_gas(Gas::from_gas(receiver_gas))
```
Comment (frol) [inline near-contract-standards/src/fungible_token/core_impl.rs:149]:
```suggestion
            .with_static_gas(receiver_gas)
```

File: near-contract-standards/src/non_fungible_token/approval/approval_impl.rs [hunk 9]
```diff
@@ -10,7 +10,7 @@ use crate::non_fungible_token::utils::{
 use crate::non_fungible_token::NonFungibleToken;
 use near_sdk::{assert_one_yocto, env, require, AccountId, Gas, Promise};
 
-const GAS_FOR_NFT_APPROVE: Gas = Gas(10_000_000_000_000);
+const GAS_FOR_NFT_APPROVE: Gas = Gas::from_ggas(10);
```
Comment (frol) [inline near-contract-standards/src/non_fungible_token/approval/approval_impl.rs]:
```suggestion
const GAS_FOR_NFT_APPROVE: Gas = Gas::from_tgas(10);
```

File: near-contract-standards/src/non_fungible_token/core/core_impl.rs [hunk 10]
```diff
@@ -15,8 +15,8 @@ use near_sdk::{
 };
 use std::collections::HashMap;
 
-const GAS_FOR_RESOLVE_TRANSFER: Gas = Gas(5_000_000_000_000);
-const GAS_FOR_NFT_TRANSFER_CALL: Gas = Gas(25_000_000_000_000 + GAS_FOR_RESOLVE_TRANSFER.0);
+const GAS_FOR_RESOLVE_TRANSFER: Gas = Gas::from_ggas(5);
+const GAS_FOR_NFT_TRANSFER_CALL: Gas = Gas::from_ggas(30);
```
Comment (frol) [inline near-contract-standards/src/non_fungible_token/core/core_impl.rs]:
```suggestion
const GAS_FOR_RESOLVE_TRANSFER: Gas = Gas::from_tgas(5);
const GAS_FOR_NFT_TRANSFER_CALL: Gas = Gas::from_tgas(30);
```

File: near-contract-standards/src/non_fungible_token/core/core_impl.rs [hunk 11]
```diff
@@ -391,13 +391,16 @@ impl NonFungibleTokenCore for NonFungibleToken {
         msg: String,
     ) -> PromiseOrValue<bool> {
         assert_one_yocto();
-        require!(env::prepaid_gas() > GAS_FOR_NFT_TRANSFER_CALL, "More gas is required");
+        require!(
+            env::prepaid_gas().as_gas() > GAS_FOR_NFT_TRANSFER_CALL.as_gas(),
```
Comment (frol) [inline near-contract-standards/src/non_fungible_token/core/core_impl.rs:395]: ^

**Review by frol — APPROVED** [review]
Looks good to me! See some proposed fixes in the comments

File: examples/fungible-token/test-contract-defi/src/lib.rs (hunk 2 above)
Comment (iho) [reply examples/fungible-token/test-contract-defi/src/lib.rs:13]: I can't do that. Those two must be consts too, but  they are not.

Comment (frol) [reply examples/fungible-token/test-contract-defi/src/lib.rs:13]: @iho What exactly is not const? `from_gas` and `as_gas` are const-fns

**Review by frol — APPROVED** [review]
I believe this PR is good to go once new version of near-gas is released.

**Review by frol — APPROVED** [review]
